Snowflake.Data/Core/QueryContextCache.cs

```diff
@@ -0,0 +1,332 @@
+
+        // When the query context read (bigint). Compare for same id.
+        public long ReadTimestamp { get; set; }
+
+        // Priority of the query context (bigint). Compare for different ids.
+        public long Priority { get; set; }
+
+        // Opaque information (object with a value of base64 encoded string).
+        public string Context { get; set; }
+
+        public QueryContextElement(long id, long timestamp,
+                            long pri, string context)
+        {
+            Id = id;
+            ReadTimestamp = timestamp;
+            Priority = pri;
+            Context = context;
+        }
+
+        public override bool Equals(Object obj)
+        {
+            if (obj == null || !(obj is QueryContextElement))
+            {
+                return false;
+            }
+
+            QueryContextElement other = (QueryContextElement)obj;
+            return (Id == other.Id) &&
+                   (Priority == other.Priority) &&
+                   (ReadTimestamp == other.ReadTimestamp) &&
+                   (Context.Equals(other.Context));
+        }
+
+        public override int GetHashCode()
+        {
+            int hash = 31;
+
+            hash = hash * 31 + (int)Id;
+            hash += (hash * 31) + (int)ReadTimestamp;
+            hash += (hash * 31) + (int)Priority;
+            hash += (hash * 31) + Context.GetHashCode();
+
+            return hash;
+        }
+    };
+
+    class ElementComparer : IComparer<QueryContextElement>
+    {
+        public int Compare(QueryContextElement x, QueryContextElement y)
+        {
+            if (x.Priority != y.Priority) return x.Priority > y.Priority ? 1 : -1;
+            if (x.Id != y.Id) return x.Id > y.Id ? 1 : -1;
+            if (x.ReadTimestamp != y.ReadTimestamp) return x.ReadTimestamp > y.ReadTimestamp ? 1 : -1;
+            return 0;
+        }
+    }
+
+    internal class QueryContextCache
+    {
+        private int _capacity; // Capacity of the cache
+        private Dictionary<long, QueryContextElement> _idMap; // Map for id and QCC
+        private Dictionary<long, QueryContextElement> _priorityMap; // Map for priority and QCC
+        private Dictionary<long, QueryContextElement> _newPriorityMap; // Intermediate map for priority and QCC for current round of merging
+        private SortedSet<QueryContextElement> _cacheSet; // Order data as per priority
+        private SFLogger _logger = SFLoggerFactory.GetLogger<QueryContextCache>();
+
+        public QueryContextCache(int capacity)
+        {
+            _capacity = capacity;
+            _idMap = new Dictionary<long, QueryContextElement>();
+            _priorityMap = new Dictionary<long, QueryContextElement>();
+            _newPriorityMap = new Dictionary<long, QueryContextElement>();
+            _cacheSet = new SortedSet<QueryContextElement>(new ElementComparer());
+        }
+
+        public void Merge(long id, long readTimestamp, long priority, string context)
```

Comment:
Do we need all of this `public`s when the class is internal?

Reply:
Make it public for test purpose (it's called in QueryContextCache.cs) but make the class internal so it won't be called from application.